src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -170,314 +153,55 @@ class FastSync(
-        processHeader(header, peer) match {
-          case Left(result)        => result
-          case Right(headerAndDif) =>
-            updateSyncState(headerAndDif._1, headerAndDif._2)
-            if (header.number == syncState.safeDownloadTarget){
-              ImportedTargetBlock
-            } else {
-              processHeaders(peer, headers.tail)
-            }
-        }
-      } else
-        HeadersProcessingFinished
-    }
-
-    private def validateHeader(header: BlockHeader, peer: Peer): Either[HeaderProcessingResult, BlockHeader] = {
-      val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
-
-      if (shouldValidate) {
-        validators.blockHeaderValidator.validate(header, blockchain.getBlockHeaderByHash) match {
-          case Right(_) =>
-            updateValidationState(header)
-            Right(header)
-
-          case Left(error) =>
-            log.warning(s"Block header validation failed during fast sync at block ${header.number}: $error")
-            Left(ValidationFailed(header, peer))
-        }
-      } else {
-        Right(header)
-      }
-    }
-
-    private def updateSyncState(header: BlockHeader, parentTd: BigInt): Unit = {
-      blockchain.save(header)
-      blockchain.save(header.hash, parentTd + header.difficulty)
-
-      if (header.number > syncState.bestBlockHeaderNumber) {
-        syncState = syncState.copy(bestBlockHeaderNumber = header.number)
-      }
-
-      syncState = syncState
-        .enqueueBlockBodies(Seq(header.hash))
-        .enqueueReceipts(Seq(header.hash))
-    }
-
-    private def updateValidationState(header: BlockHeader): Unit ={
-      import syncConfig.{fastSyncBlockValidationK => K, fastSyncBlockValidationX => X}
-      syncState = syncState.updateNextBlockToValidate(header, K, X)
+      case InitTargetBlockUpdate(state) => handleStateChange(syncState)(fastSyncStateHandler.updateTargetBlock(state))
     }
 
-    private def processHeader(header: BlockHeader, peer: Peer) : Either[HeaderProcessingResult , (BlockHeader, BigInt)] = for {
-      validatedHeader  <- validateHeader(header, peer)
-      parentDifficulty <- getParentDifficulty(header)
-    } yield (validatedHeader, parentDifficulty)
-
-    private def getParentDifficulty(header: BlockHeader) = {
-      blockchain.getTotalDifficultyByHash(header.parentHash).toRight(ParentDifficultyNotFound(header))
-    }
-
-    private def handleBlockValidationError(header: BlockHeader, peer: Peer, N: Int): Unit = {
-      blacklist(peer.id, blacklistDuration, "block header validation failed")
-      if (header.number <= syncState.safeDownloadTarget) {
-        discardLastBlocks(header.number, N)
-        syncState = syncState.updateDiscardedBlocks(header, N)
-        if (header.number >= syncState.targetBlock.number) {
-          updateTargetBlock(LastBlockValidationFailed)
-        } else {
+    def handleStateChange(currentState: SyncState)(f: SyncState => (FastSyncCommand, SyncState)): Unit ={
```

Comment:
This is quite nice but maybe fsm could be more suitable here? We have a lot of state changes and is pretty hard to  understand the flow.

Reply:
Yes. I used akka fsm in my previous job and code looked much cleaner.